README.md
# H5UnityLoader
A simple wrapper for PInvoke for loading H5 files in Unity

Directly supports loading:

* 1D arrays of: `float`, `int`, `string`
* 2D arrays of: `float`, `int`
* Advanced: Also supports generic methods to load any datatype, but these generic types do not play well with unity, and must usually be cast manually to unity-friendly types. They are typically types `byte`, `ulong`, `double`, etc.

Tested on macos and windows 64 bit. Should also work on linux but untested.

If h5 is from python, it's easier (especially for strings) if datasets are created from numpy arrays with the dtype set to `'i8'` (int), `'float'`, or `'S'` (string).

## Installation
 
#### Unity 2020.1 and later (recommended):
 
1. Edit > Project settings > Package manager
2. Add a new scoped registry fill in the following information:
    ```text
    name: AdamBebko
    url: https://registry.npmjs.org
    scope: com.adambebko.unityh5loader
    ```
3. Go to Window > Package Manager > open the "My Registries tab"
4. Click on the package and hit install.
 
You can use this to update the package easily in the future.
 
#### Unity 2019 (not recommended):
 
From git url in package manager: 
 ```text
https://github.com/AdamBebko/UnityH5Loader.git#X.Y.Z
```

Where x.y.z is the most recent release as seen on the [GitHub](https://github.com/AdamBebko/UnityH5Loader) page.

It is much harder to get updates using this method.

#### As old .unitypackage (not recommended at all):
 
go to the [GitHub](https://github.com/BioMotionLab/UnityPDFDisplay) page and download the contents of the "upm" branch as a zip. Drag unzipped folder into assets folder.

## Use:

reference the H5Loader assembly and namespace.

There are public static methods for each kind of supported datatype. For example:

```c#
using UnityH5Loader;

float[] loadedFloatArray = H5Loader.LoadFloatDataset(filePath, "floatdatasetname");
int[,] loadedInt2DArray = H5Loader.Load2dIntDataset(filePath, "Dtwointdatasetname");
```

See included sample (accessible in package manager window) for more examples.

CHANGELOG.md
## [1.0.1](https://github.com/AdamBebko/UnityH5Loader/compare/release/1.0.0...release/1.0.1) (2020-09-29)


### Bug Fixes

* licence and readme, cleaned up unused files ([a01c5ac](https://github.com/AdamBebko/UnityH5Loader/commit/a01c5ac11bd006dd2d750e5bda8d06cec6ffb2a6))

# 1.0.0 (2020-09-29)


### Bug Fixes

* first release ([a6b343b](https://github.com/AdamBebko/UnityH5Loader/commit/a6b343be60049840d6a55c0f684f47dcb79c25f1))
* save scene ([4f5abf0](https://github.com/AdamBebko/UnityH5Loader/commit/4f5abf0690d972cc050b448db6704f4da198df88))


### Features

* Getting ready for release ([aa42974](https://github.com/AdamBebko/UnityH5Loader/commit/aa4297489a46d5d054914d314b83fad865ab494b))


Scripts/H5Loader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using HDF.PInvoke;
using JetBrains.Annotations;
using UnityEngine;

namespace H5Loader {
    public static class H5Loader {

        static readonly ulong[] MaxDimensions = {10000, 10000, 10000};

        /// <summary>
        /// Loads a 1D int dataset from an H5 file. The dataset must be 'i8' encoded dataset. In python: create it using numpy array with dtype='i8'
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <returns></returns>
        [PublicAPI]
        public static int[] LoadIntDataset(string filePath, string datasetName) {
            long[] longArray = LoadDataset<long>(filePath, datasetName);
            int[] integerArray = longArray.Select(item => (int) item).ToArray();
            return integerArray;
        }

        /// <summary>
        /// Loads a 1d float dataset from an H5 file. The dataset must be 'float' encoded dataset. In python: create it using numpy array with dtype='float'
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <returns></returns>
        [PublicAPI]
        public static float[] LoadFloatDataset(string filePath, string datasetName) {
            double[] doubleArray = LoadDataset<double>(filePath, datasetName);
            float[] floatArray = doubleArray.Select(item => (float) item).ToArray();
            return floatArray;
        }

        /// <summary>
        /// Loads a 1d string dataset from an H5 file. The dataset must be 'string' encoded dataset. In python: create it using numpy array with dtype='S'
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="dataSetName"></param>
        /// <returns></returns>
        [PublicAPI]
        public static string[] LoadStringDataset(string filePath, string dataSetName) {
            //With much Help from:  https://stackoverflow.com/questions/23295545/reading-string-array-from-a-hdf5-dataset
            long fileId = H5F.open(filePath, H5F.ACC_RDONLY);
            
            string longJoinedString;
            int stringLength;
            try {
                
                long datasetId = H5D.open(fileId, dataSetName);
                long spaceID = H5D.get_space(datasetId);
                long dataType = H5D.get_type(datasetId);

                int[] dimensions = GetDatasetDimensions(spaceID);

                stringLength = (int) H5T.get_size(dataType);
                byte[] buffer = new byte[dimensions[0] * stringLength];

                GCHandle gch = GCHandle.Alloc(buffer, GCHandleType.Pinned);

                
                try {
                    H5D.read(datasetId, dataType, H5S.ALL, H5S.ALL, H5P.DEFAULT, gch.AddrOfPinnedObject());
                    longJoinedString = Encoding.ASCII.GetString(buffer);
                }
                finally {
                    gch.Free();
                    H5D.close(dataType);
                    H5D.close(spaceID);
                    H5D.close(datasetId);
                }
            }
            finally {
                H5F.close(fileId);
            }

            return longJoinedString.SplitInParts(stringLength).Select(ss => (string) (object) ss).ToArray();
        }

        static IEnumerable<string> SplitInParts(this string theLongString, int partLength) {
            if (theLongString == null)
                throw new ArgumentNullException(nameof(theLongString));
            if (partLength <= 0)
                throw new ArgumentException("Part length has to be positive.", nameof(partLength));

            for (var i = 0; i < theLongString.Length; i += partLength)
                yield return theLongString.Substring(i, Math.Min(partLength, theLongString.Length - i));
        }

        /// <summary>
        /// WARNING: ADVANCED USE ONLY!! Loads a 1D generic dataset from an H5 file.
        /// The generic loaders only loads data in non-Unity friendly types, such as bytes, uints, longs etc...
        /// You'll have to know the correct cast to retrieve usable data.
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <typeparam name="T"></typeparam>
        /// <returns></returns>
        /// <exception cref="FileNotFoundException"></exception>
        [PublicAPI]
        public static T[] LoadDataset<T>(string filePath, string datasetName) {
            
            if (!File.Exists(filePath)) throw new FileNotFoundException($"Loading dataset {datasetName} from file that doesn't exist {filePath}");
            long fileId = H5F.open(filePath, H5F.ACC_RDONLY);
            
            T[] resultArray;

            try {

                long datasetId = H5D.open(fileId, datasetName);
                if (datasetId == -1) throw new ArgumentException($"Dataset could not be opened. Check filepath exists and is correct. FilePath = {filePath}");
                long typeId = H5D.get_type(datasetId);
                long spaceID = H5D.get_space(datasetId);

                int[] dimensions = GetDatasetDimensions(spaceID);

                resultArray = new T[dimensions[0]];
                GCHandle gch = GCHandle.Alloc(resultArray, GCHandleType.Pinned);

                try {
                    H5D.read(datasetId, typeId, H5S.ALL, H5S.ALL, H5P.DEFAULT, gch.AddrOfPinnedObject());
                }
                finally {
                    gch.Free();
                    H5D.close(typeId);
                    H5D.close(spaceID);
                    H5D.close(datasetId);
                }
            }
            finally {
                H5F.close(fileId);
            }

            return resultArray;

        }
        
        /// <summary>
        /// WARNING: ADVANCED USE ONLY!! Loads a 2D generic dataset from an H5 file.
        /// The generic loaders only loads data in non-Unity friendly types, such as bytes, uints, longs etc...
        /// You'll have to know the correct cast to retrieve usable data.
        /// 
        /// Created With help from https://github.com/LiorBanai/HDF5-CSharp/blob/master/HDF5-CSharp/Hdf5Dataset.cs
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <typeparam name="T"></typeparam>
        /// <returns></returns>
        /// <exception cref="FileNotFoundException"></exception>
        static T[,] Load2DDataset<T>(string filePath, string datasetName) {
            
            if (!File.Exists(filePath)) throw new FileNotFoundException($"Loading dataset {datasetName} from file that doesn't exist {filePath}");
            long fileId = H5F.open(filePath, H5F.ACC_RDONLY);
            
            T[,] resultArray = new T[2,2];
            try {
                
                ulong[] start = {0, 0};
                ulong[] count = {0, 0};

                long datasetId = H5D.open(fileId, datasetName);
                var datatype = H5D.get_type(datasetId);
                var spaceId = H5D.get_space(datasetId);
                int rank = H5S.get_simple_extent_ndims(spaceId);
                ulong[] maxDims = new ulong[rank];
                ulong[] dims = new ulong[rank];
                H5S.get_simple_extent_dims(spaceId, dims, maxDims);
                
                count[0] = dims[0];
                count[1] = dims[1];

                // Define file hyperslab. 
                long status = H5S.select_hyperslab(spaceId, H5S.seloper_t.SET, start, null, count, null);
                
                // Define the memory dataspace.
                resultArray = new T[dims[0], dims[1]];
                var memId = H5S.create_simple(rank, dims, null);

                // Define memory hyperslab. 
                status = H5S.select_hyperslab(memId, H5S.seloper_t.SET, start, null,
                    count, null);

                // Read data from hyperslab in the file into the hyperslab in 
                // memory and display.             
                GCHandle handle = GCHandle.Alloc(resultArray, GCHandleType.Pinned);

                try {
                    H5D.read(datasetId, datatype, memId, spaceId,
                        H5P.DEFAULT, handle.AddrOfPinnedObject());
                }
                finally {
                    handle.Free();
                    H5S.close(status);
                    H5S.close(memId);
                    H5S.close(spaceId);
                    H5D.close(datatype);
                    H5D.close(datasetId);
                }
            }
            finally {
                H5F.close(fileId);
            }
            return resultArray;

        }
        

    
        static int[] GetDatasetDimensions(long spaceID) {
            int numberOfDimensions = H5S.get_simple_extent_ndims(spaceID);

            int[] dimensions = new int[0];
            if (numberOfDimensions >= 0) {
                ulong[] dims = new ulong[numberOfDimensions];
                H5S.get_simple_extent_dims(spaceID, dims, MaxDimensions);
                dimensions = ConvertDimensionsToIntegers(dims);
            }

            return dimensions;
        }
        
        static int[] ConvertDimensionsToIntegers(ulong[] dims) {
            if (dims == null) throw new ArgumentNullException(nameof(dims));
            int[] dimensions = new int[dims.Length];
            for (int i = 0; i < dims.Length; i++) {
                dimensions[i] = (int) dims[i];
            }

            return dimensions;
        }

        /// <summary>
        /// Loads a 2d integer dataset from an H5 file. The dataset must be an 'i8' encoded dataset. In python: create it using numpy array with dtype='i8'
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <returns></returns>
        [PublicAPI]
        public static int[,] Load2dIntDataset(string filePath, string datasetName) {
            long[,] doubleArray = Load2DDataset<long>(filePath, datasetName);
            int[,] intArray = new int[doubleArray.GetLength(0),doubleArray.GetLength(1)];
            for (int i = 0; i < doubleArray.GetLength(0); i++) {
                for (int j = 0; j < doubleArray.GetLength(1); j++) {
                    intArray[i, j] = (int)doubleArray[i, j];
                }
            }
            return intArray;
        }
        
        /// <summary>
        /// Loads a 2d float dataset from an H5 file. The dataset must be 'float' encoded dataset. In python: create it using numpy array with dtype='float'
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <returns></returns>
        public static float[,] Load2dFloatDataset(string filePath, string datasetName) {
            double[,] doubleArray = Load2DDataset<double>(filePath, datasetName);
            float[,] floatArray = new float[doubleArray.GetLength(0),doubleArray.GetLength(1)];
            for (int i = 0; i < doubleArray.GetLength(0); i++) {
                for (int j = 0; j < doubleArray.GetLength(1); j++) {
                    floatArray[i, j] = (float)doubleArray[i, j];
                }
            }
            return floatArray;
        }
        
    }

}


H5Loader_Project/Assets/Scripts/UnityH5.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using HDF.PInvoke;
using JetBrains.Annotations;
using UnityEngine;

namespace MoshPlayer.Scripts.FileLoaders {
    public static class UnityH5 {

        static readonly ulong[] MaxDimensions = {10000, 10000, 10000};

        /// <summary>
        /// Loads a 1D int dataset from an H5 file. The dataset must be 'i8' encoded dataset. In python: create it using numpy array with dtype='i8'
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <returns></returns>
        [PublicAPI]
        public static int[] LoadIntDataset(string filePath, string datasetName) {
            long[] longArray = LoadDataset<long>(filePath, datasetName);
            int[] integerArray = longArray.Select(item => (int) item).ToArray();
            return integerArray;
        }

        /// <summary>
        /// Loads a 1d float dataset from an H5 file. The dataset must be 'float' encoded dataset. In python: create it using numpy array with dtype='float'
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <returns></returns>
        [PublicAPI]
        public static float[] LoadFloatDataset(string filePath, string datasetName) {
            double[] doubleArray = LoadDataset<double>(filePath, datasetName);
            float[] floatArray = doubleArray.Select(item => (float) item).ToArray();
            return floatArray;
        }

        /// <summary>
        /// Loads a 1d string dataset from an H5 file. The dataset must be 'string' encoded dataset. In python: create it using numpy array with dtype='S'
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="dataSetName"></param>
        /// <returns></returns>
        [PublicAPI]
        public static string[] LoadStringDataset(string filePath, string dataSetName) {
            //With much Help from:  https://stackoverflow.com/questions/23295545/reading-string-array-from-a-hdf5-dataset
            long fileId = H5F.open(filePath, H5F.ACC_RDONLY);
            
            string longJoinedString;
            int stringLength;
            try {
                
                long datasetId = H5D.open(fileId, dataSetName);
                long spaceID = H5D.get_space(datasetId);
                long dataType = H5D.get_type(datasetId);

                int[] dimensions = GetDatasetDimensions(spaceID);

                stringLength = (int) H5T.get_size(dataType);
                byte[] buffer = new byte[dimensions[0] * stringLength];

                GCHandle gch = GCHandle.Alloc(buffer, GCHandleType.Pinned);

                
                try {
                    H5D.read(datasetId, dataType, H5S.ALL, H5S.ALL, H5P.DEFAULT, gch.AddrOfPinnedObject());
                    longJoinedString = Encoding.ASCII.GetString(buffer);
                }
                finally {
                    gch.Free();
                    H5D.close(dataType);
                    H5D.close(spaceID);
                    H5D.close(datasetId);
                }
            }
            finally {
                H5F.close(fileId);
            }

            return longJoinedString.SplitInParts(stringLength).Select(ss => (string) (object) ss).ToArray();
        }

        static IEnumerable<string> SplitInParts(this string theLongString, int partLength) {
            if (theLongString == null)
                throw new ArgumentNullException(nameof(theLongString));
            if (partLength <= 0)
                throw new ArgumentException("Part length has to be positive.", nameof(partLength));

            for (var i = 0; i < theLongString.Length; i += partLength)
                yield return theLongString.Substring(i, Math.Min(partLength, theLongString.Length - i));
        }

        /// <summary>
        /// WARNING: ADVANCED USE ONLY!! Loads a 1D generic dataset from an H5 file.
        /// The generic loaders only loads data in non-Unity friendly types, such as bytes, uints, longs etc...
        /// You'll have to know the correct cast to retrieve usable data.
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <typeparam name="T"></typeparam>
        /// <returns></returns>
        /// <exception cref="FileNotFoundException"></exception>
        [PublicAPI]
        public static T[] LoadDataset<T>(string filePath, string datasetName) {
            
            if (!File.Exists(filePath)) throw new FileNotFoundException($"Loading dataset {datasetName} from file that doesn't exist {filePath}");
            long fileId = H5F.open(filePath, H5F.ACC_RDONLY);
            
            T[] resultArray;

            try {

                long datasetId = H5D.open(fileId, datasetName);
                if (datasetId == -1)
                    Debug.LogError("Dataset could not be opened. Check filepath exists and is correct");
                long typeId = H5D.get_type(datasetId);
                long spaceID = H5D.get_space(datasetId);

                int[] dimensions = GetDatasetDimensions(spaceID);

                resultArray = new T[dimensions[0]];
                GCHandle gch = GCHandle.Alloc(resultArray, GCHandleType.Pinned);

                try {
                    H5D.read(datasetId, typeId, H5S.ALL, H5S.ALL, H5P.DEFAULT, gch.AddrOfPinnedObject());
                }
                finally {
                    gch.Free();
                    H5D.close(typeId);
                    H5D.close(spaceID);
                    H5D.close(datasetId);
                }
            }
            finally {
                H5F.close(fileId);
            }

            return resultArray;

        }
        
        /// <summary>
        /// WARNING: ADVANCED USE ONLY!! Loads a 2D generic dataset from an H5 file.
        /// The generic loaders only loads data in non-Unity friendly types, such as bytes, uints, longs etc...
        /// You'll have to know the correct cast to retrieve usable data.
        /// 
        /// Created With help from https://github.com/LiorBanai/HDF5-CSharp/blob/master/HDF5-CSharp/Hdf5Dataset.cs
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <typeparam name="T"></typeparam>
        /// <returns></returns>
        /// <exception cref="FileNotFoundException"></exception>
        static T[,] Load2DDataset<T>(string filePath, string datasetName) {
            
            if (!File.Exists(filePath)) throw new FileNotFoundException($"Loading dataset {datasetName} from file that doesn't exist {filePath}");
            long fileId = H5F.open(filePath, H5F.ACC_RDONLY);
            
            T[,] resultArray = new T[2,2];
            try {
                
                ulong[] start = {0, 0};
                ulong[] count = {0, 0};

                long datasetId = H5D.open(fileId, datasetName);
                var datatype = H5D.get_type(datasetId);
                var spaceId = H5D.get_space(datasetId);
                int rank = H5S.get_simple_extent_ndims(spaceId);
                ulong[] maxDims = new ulong[rank];
                ulong[] dims = new ulong[rank];
                H5S.get_simple_extent_dims(spaceId, dims, maxDims);
                
                count[0] = dims[0];
                count[1] = dims[1];

                // Define file hyperslab. 
                long status = H5S.select_hyperslab(spaceId, H5S.seloper_t.SET, start, null, count, null);
                
                // Define the memory dataspace.
                resultArray = new T[dims[0], dims[1]];
                var memId = H5S.create_simple(rank, dims, null);

                // Define memory hyperslab. 
                status = H5S.select_hyperslab(memId, H5S.seloper_t.SET, start, null,
                    count, null);

                // Read data from hyperslab in the file into the hyperslab in 
                // memory and display.             
                GCHandle handle = GCHandle.Alloc(resultArray, GCHandleType.Pinned);

                try {
                    H5D.read(datasetId, datatype, memId, spaceId,
                        H5P.DEFAULT, handle.AddrOfPinnedObject());
                }
                finally {
                    handle.Free();
                    H5S.close(status);
                    H5S.close(memId);
                    H5S.close(spaceId);
                    H5D.close(datatype);
                    H5D.close(datasetId);
                }
            }
            finally {
                H5F.close(fileId);
            }
            return resultArray;

        }
        

    
        static int[] GetDatasetDimensions(long spaceID) {
            int numberOfDimensions = H5S.get_simple_extent_ndims(spaceID);

            int[] dimensions = new int[0];
            if (numberOfDimensions >= 0) {
                ulong[] dims = new ulong[numberOfDimensions];
                H5S.get_simple_extent_dims(spaceID, dims, MaxDimensions);
                dimensions = ConvertDimensionsToIntegers(dims);
            }

            return dimensions;
        }
        
        static int[] ConvertDimensionsToIntegers(ulong[] dims) {
            if (dims == null) throw new ArgumentNullException(nameof(dims));
            int[] dimensions = new int[dims.Length];
            for (int i = 0; i < dims.Length; i++) {
                dimensions[i] = (int) dims[i];
            }

            return dimensions;
        }

        /// <summary>
        /// Loads a 2d integer dataset from an H5 file. The dataset must be an 'i8' encoded dataset. In python: create it using numpy array with dtype='i8'
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <returns></returns>
        [PublicAPI]
        public static int[,] Load2dIntDataset(string filePath, string datasetName) {
            long[,] doubleArray = Load2DDataset<long>(filePath, datasetName);
            int[,] intArray = new int[doubleArray.GetLength(0),doubleArray.GetLength(1)];
            for (int i = 0; i < doubleArray.GetLength(0); i++) {
                for (int j = 0; j < doubleArray.GetLength(1); j++) {
                    intArray[i, j] = (int)doubleArray[i, j];
                }
            }
            return intArray;
        }
        
        /// <summary>
        /// Loads a 2d float dataset from an H5 file. The dataset must be 'float' encoded dataset. In python: create it using numpy array with dtype='float'
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="datasetName"></param>
        /// <returns></returns>
        public static float[,] Load2dFloatDataset(string filePath, string datasetName) {
            double[,] doubleArray = Load2DDataset<double>(filePath, datasetName);
            float[,] floatArray = new float[doubleArray.GetLength(0),doubleArray.GetLength(1)];
            for (int i = 0; i < doubleArray.GetLength(0); i++) {
                for (int j = 0; j < doubleArray.GetLength(1); j++) {
                    floatArray[i, j] = (float)doubleArray[i, j];
                }
            }
            return floatArray;
        }
        
    }

}
